### PR TITLE
Remove all C style casts and enable compile error for them

### LINKED ISSUE #675: Remove all C style casts
From discussion with @artemp we should replace all C style casts with the correct C++ cast.

It may be possible to use g++'s `-Wold-style-cast` flag to find them. Clang might also support this too.

### PR DESCRIPTION
**Pull request by ljbade**
Closes #675 

@jfirebaugh Can you review?

### CODE REVIEW

**Comment by ljbade** [comment]
Also depends on https://github.com/mapbox/geojson-vt-cpp/tree/no-c-cast merge (which in turn depends on https://github.com/mapbox/geojson-vt-cpp/tree/no-std_to_string which is not merged into it's `master` yet is currently referred from this `master`)

**Comment by ljbade** [comment]
According to @kkaefer `geojson-vt` `master` is not ready for merging yet into `mapbox-gl-native`. So I will leave it pegged to my branch which is based on the current branch from @kkaefer.

File: platform/ios/MGLMultiPoint.mm [hunk 1]
```diff
@@ -18,7 +18,7 @@ - (instancetype)initWithCoordinates:(CLLocationCoordinate2D *)coords
     if (self)
     {
         _count = count;
-        _coords = (CLLocationCoordinate2D *)malloc(_count * sizeof(CLLocationCoordinate2D));
+        _coords = reinterpret_cast<CLLocationCoordinate2D *>(malloc(_count * sizeof(CLLocationCoordinate2D)));
```
Comment (jfirebaugh) [inline platform/ios/MGLMultiPoint.mm]: This should be using `new []`, not `malloc`.

File: src/clipper/clipper.cpp [hunk 2]
```diff
@@ -4334,10 +4335,10 @@ double DistanceFromLineSqrd(
   const IntPoint& pt, const IntPoint& ln1, const IntPoint& ln2)
 {
   //The equation of a line in general form (Ax + By + C = 0)
-  //given 2 points (x�,y�) & (x�,y�) is ...
-  //(y� - y�)x + (x� - x�)y + (y� - y�)x� - (x� - x�)y� = 0
-  //A = (y� - y�); B = (x� - x�); C = (y� - y�)x� - (x� - x�)y�
-  //perpendicular distance of point (x�,y�) = (Ax� + By� + C)/Sqrt(A� + B�)
+  //given 2 points (x�,y�) & (x�,y�) is ...
```
Comment (jfirebaugh) [inline src/clipper/clipper.cpp]: Looks like your text editor corrupted this file.

File: src/clipper/clipper.cpp [hunk 3]
```diff
@@ -4601,7 +4602,7 @@ void OffsetPaths(const Paths &in_polys, Paths &out_polys,
   double delta, JoinType jointype, EndType_ endtype, double limit)
 {
   ClipperOffset co(limit, limit);
-  co.AddPaths(in_polys, jointype, (EndType)endtype); 
+  co.AddPaths(in_polys, jointype, (EndType)endtype);
```
Comment (jfirebaugh) [inline src/clipper/clipper.cpp]: Is this file entirely whitespace changes? Can you change your editor settings to modify whitespace only on lines that actually changed, or separate whitespace-only changes into a separate commit?

File: src/rapidjson/reader.h [hunk 4]
```diff
@@ -175,7 +175,7 @@ inline const char *SkipWhitespace_SIMD(const char* p) {
 
 #ifdef RAPIDJSON_SIMD
 //! Template function specialization for InsituStringStream
-template<> inline void SkipWhitespace(InsituStringStream& stream) { 
+template<> inline void SkipWhitespace(InsituStringStream& stream) {
```
Comment (jfirebaugh) [inline src/rapidjson/reader.h]: Whitespace only changes; 3rd party file -- please revert.

**Comment by jfirebaugh** [comment]
Cast changes look :+1:.

**Comment by ljbade** [comment]
Ah sorry @jfirebaugh my editor has delete whitespace at end of lines enabled. Let me fix this up manually with `nano`.

**Comment by incanus** [comment]
This is problematic in Objective-C++, where it's quite common to C-style cast.

**Comment by jfirebaugh** [comment]
It also triggered compile errors in rapidjson.h. I've force-pushed master back to d4159d51419ae59824c1a20cb682d2a66150a7ea, prior to this change.

**Comment by ljbade** [comment]
@jfirebaugh Where were you seeing `rapidjson.h` errors? I thought I had isolated all of those using the push/pop #defines

**Comment by jfirebaugh** [comment]
Not seeing the rapidjson.h errors on the 675-no-c-cast branch but here are some of the errors when biulding the osxapp target in Xcode:

```
/Users/john/Development/mapbox-gl-native/platform/darwin/settings_nsuserdefaults.mm:35:28: error: use of old-style cast [-Werror,-Wold-style-cast]
        userTrackingMode = (MGLUserTrackingMode)uncheckedTrackingMode;
```

```
In file included from /Users/john/Development/mapbox-gl-native/macosx/main.mm:20:
In file included from /Users/john/Development/mapbox-gl-native/mason_packages/headers/boost/1.57.0/include/boost/program_options.hpp:15:
In file included from /Users/john/Development/mapbox-gl-native/mason_packages/headers/boost/1.57.0/include/boost/program_options/options_description.hpp:13:
In file included from /Users/john/Development/mapbox-gl-native/mason_packages/headers/boost/1.57.0/include/boost/program_options/value_semantic.hpp:13:
In file included from /Users/john/Development/mapbox-gl-native/mason_packages/headers/boost/1.57.0/include/boost/function/function1.hpp:11:
In file included from /Users/john/Development/mapbox-gl-native/mason_packages/headers/boost/1.57.0/include/boost/function/detail/maybe_include.hpp:18:
/Users/john/Development/mapbox-gl-native/mason_packages/headers/boost/1.57.0/include/boost/function/function_template.hpp:628:37: error: use of old-style cast [-Werror,-Wold-style-cast]
          functor.obj_ref.obj_ptr = (void *)(f.get_pointer());
                                    ^       ~~~~~~~~~~~~~~~~~
```

I guess this target isn't build by CI, which is why it wasn't caught.

**Comment by incanus** [comment]
> I guess this target isn't build by CI

Hmm, does https://travis-ci.org/mapbox/mapbox-gl-native/jobs/75360573#L77 not cover it? Should we add another entry to the matrix? :disappointed:

**Comment by jfirebaugh** [comment]
Nice @incanus -- from that very build:

```
⌦  /Users/travis/build/mapbox/mapbox-gl-native/platform/darwin/settings_nsuserdefaults.mm:35:28: use of old-style cast [-Werror,-Wold-style-cast]
        userTrackingMode = (MGLUserTrackingMode)uncheckedTrackingMode;
                           ^                    ~~~~~~~~~~~~~~~~~~~~~
** BUILD FAILED **
```

But for some reason this didn't fail the overall build.

**Comment by ljbade** [comment]
@jfirebaugh that is very bizzare. No wonder I didn't pick it up. I really should dust off my Mac mini more often for testing my branches instead of relying on Travis.

**Pull request merged**